### PR TITLE
webgpu: Implement device lost

### PR DESCRIPTION
**Pull request by sagudev**
Due to hitting multiple deadlocks in wgpu I added lock to prevent some functions (queue_write_texture, queue_submit) run in parallel with device_poll in poller thread https://github.com/servo/servo/pull/32354/commits/d9d506c6844dc41e1a43363804e6d0445b224ca0 (and thus avoiding deadlocks).

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix https://github.com/servo/servo/issues/32347
- [x] There are tests for these changes in CTS

### CODE REVIEW

**Comment by sagudev** [comment]
Current try run: https://github.com/sagudev/servo/actions/runs/9205601757/job/25323173909

**Comment by sagudev** [comment]
I think I will not deal with `_webgpu/webgpu/cts.https.html?q=webgpu:api,operation,adapter,requestDevice:invalid:*` in this PR as wgpu-core allows creating multiple valid devices from same adapter, but spec doesn't.

**Comment by sagudev** [comment]
new run: https://github.com/sagudev/servo/actions/runs/9211814858/job/25343380939
`_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,state,device_lost,destroy` has (only in CI):
```
PASS [expected TIMEOUT] subtest: :stage="finish";awaitLost=true
PASS [expected NOTRUN] subtest: :stage="finish";awaitLost=false
PASS [expected NOTRUN] subtest: :stage="submit";awaitLost=true
FAIL [expected NOTRUN] subtest: :stage="submit";awaitLost=false

assert_unreached: 
  - EXCEPTION: endTestScope timed out
    PromiseTimeoutError@https://web-platform.test:8443/_webgpu/webgpu/common/util/util.js:114:8
    raceWithRejectOnTimeout/timeoutPromise&lt;/handle&lt;@https://web-platform.test:8443/_webgpu/webgpu/common/util/util.js:140:14
    step_timeout/&lt;@https://web-platform.test:8443/resources/testharness.js:1234:18
    
 Reached unreachable code
```

`_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,state,device_lost,destroy:command,renderPass,renderBundle:*` is crashing now (instead of timeout), but it's out of scope of this PR.

**Comment by sagudev** [comment]
`endTestScope timed out` are happening because WGPU thread is stuck in deadlock: https://github.com/gfx-rs/wgpu/issues/5737

**Comment by sagudev** [comment]
https://github.com/sagudev/servo/actions/runs/9271180193/job/25507347414

`_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,state,device_lost,destroy:createComputePipelineAsync:*` and `_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,state,device_lost,destroy:createRenderPipelineAsync:*` is due to https://github.com/servo/servo/issues/31591

`_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,state,device_lost,destroy:queue,writeTexture,2d,uncompressed_format:*` has flaky timeout that I have yet to investigate.

**Comment by sagudev** [comment]
> 
> `_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,state,device_lost,destroy:queue,writeTexture,2d,uncompressed_format:*` has flaky timeout that I have yet to investigate.

This is due to another deadlock issue in wgpu: https://github.com/gfx-rs/wgpu/issues/5737#issuecomment-2163556055

**Comment by gterzian** [comment]
Will review tomorrow!

**Review by gterzian — CHANGES_REQUESTED** [review]
LGTM, but I don't think we need to propagate all the lost reasons.

File: components/webgpu/script_messages.rs [hunk 1]
```diff
@@ -14,6 +14,58 @@ use crate::wgc::id::{
...
 };
+use crate::wgt;
+
+// Workaround until https://github.com/gfx-rs/wgpu/pull/5732
+#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
+pub enum DeviceLostReason {
+    /// Triggered by driver
+    Unknown = 0,
+    /// After Device::destroy
+    Destroyed = 1,
```
[the remark this answers is not in the document]
Comment (sagudev) [reply components/webgpu/script_messages.rs:25]: I just copied this from wgpu as I planned to use wgpu's type, but we could definitely roll our own type (that would not have other irrelevant fields). That would mean that we will do 2 conversions: wgt::DeviceLostReason -> DeviceLostReason -> GPUDeviceLostReason

File: components/script/dom/gpudevice.rs [hunk 2]
```diff
@@ -206,18 +217,11 @@ impl GPUDevice {
...
-                    "Device self-destruction sequence activated successfully!"
-                },
-            };
-            let lost = GPUDeviceLostInfo::new(global, msg.into(), reason);
-            lost_promise.resolve_native(&*lost);
-        }
+    pub fn lose(&self, reason: GPUDeviceLostReason, msg: String) {
+        let ref lost_promise = *self.lost_promise.borrow();
+        let global = &self.global();
+        let lost = GPUDeviceLostInfo::new(global, msg.into(), reason);
```
[the remark this answers is not in the document]
Comment (sagudev) [reply components/script/dom/gpudevice.rs:223]: Given the experimental nature of servo (and even more experimental WebGPU support) I chose more (servo/wgpu) developer friendly approach by propagate wgpu error messages.

File: components/script/dom/gpudevice.rs [hunk 3]
```diff
@@ -100,6 +101,7 @@ impl GPUDevice {
         device: webgpu::WebGPUDevice,
         queue: &GPUQueue,
         label: String,
+        lost_promise: Rc<Promise>,
```
[the remark this answers is not in the document]
Comment (sagudev) [reply components/script/dom/gpudevice.rs:104]: I am not sure exactly what you meant here, but this promise is created in `new`, that calls `new_inherited`.

**Comment by sagudev** [comment]
Lifetime of `GPUDevice` still bugs me, global scope has hashmap of `Dom<GPUDevice>` that keeps `GPUDevice` alive unless GPUDevice is manually destroyed, so we should probably use weakref in hashmap: https://doc.servo.org/script/dom/bindings/weakref/index.html, but this can be done as a follow up (this problem exist also in main).

File: components/script/dom/gpudevice.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gterzian) [reply components/script/dom/gpudevice.rs:104]: Ah yes, ok.

**Review by gterzian — APPROVED** [review]
LGTM. 

>we should probably use weakref in hashmap

Add a note before merging?

**Comment by sagudev** [comment]
> > we should probably use weakref in hashmap
> 
> Add a note before merging?

I opened https://github.com/servo/servo/issues/32519 and assigned myself as I already have WIP.

**Pull request merged**